src/output-json.c

```diff
@@ -77,6 +77,82 @@
 
 #define MAX_JSON_SIZE 2048
 
+#ifdef HAVE_GEOIP
+#include <maxminddb.h>
+
+static MMDB_s mmdb;
+static int mmdb_status = MMDB_FILE_OPEN_ERROR;
+
+void GeoIPGet(JsonBuilder *js, const MMDB_s *const mmdb, const char *ip_address, const char *key);
+
+#define GeoIPSetString(js, entry_data, key)                                                        \
+    {                                                                                              \
+        if (entry_data.has_data && entry_data.utf8_string != NULL) {                               \
+            jb_set_string_from_bytes(                                                              \
+                    js, key, (const uint8_t *)entry_data.utf8_string, entry_data.data_size);       \
+        }                                                                                          \
+    }
+
+void GeoIPGet(JsonBuilder *js, const MMDB_s *const mmdb, const char *ip_address, const char *key)
+{
+    int gai_error, mmdb_error;
+    MMDB_entry_data_s entry_data;
+
+    if (mmdb_status != MMDB_SUCCESS) {
+        return;
+    }
+    MMDB_lookup_result_s result = MMDB_lookup_string(mmdb, ip_address, &gai_error, &mmdb_error);
+    if (MMDB_SUCCESS != gai_error) {
+        return;
+    }
+
+    if (!result.found_entry) {
+        return;
+    }
+
+    jb_open_object(js, key);
+    jb_set_string(js, "ip", ip_address);
+
+    /* Create geo object */
+    jb_open_object(js, "geo");
+
+    if (MMDB_get_value(&result.entry, &entry_data, "continent", "code", NULL) == MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "continent_code");
+    }
+    if (MMDB_get_value(&result.entry, &entry_data, "country", "iso_code", NULL) == MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "country_iso_code");
+    }
+    if (MMDB_get_value(&result.entry, &entry_data, "city", "names", "en", NULL) == MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "city_name");
+    }
+    if (MMDB_get_value(&result.entry, &entry_data, "country", "names", "en", NULL) ==
+            MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "country_name");
+    }
+    if (MMDB_get_value(&result.entry, &entry_data, "continent", "names", "en", NULL) ==
+            MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "continent_name");
+    }
+    if (MMDB_get_value(&result.entry, &entry_data, "location", "time_zone", NULL) == MMDB_SUCCESS) {
+        GeoIPSetString(js, entry_data, "timezone");
+    }
+
+    /* Create location object */
+    jb_open_object(js, "location");
```

Comment:
this object would be empty if the 2 calls for location don't yield data. So it's best to do these calls before opening the object, and only open the object if we have something to log